File: leetcode/middleNode.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def middleNode(self, head: ListNode) -> ListNode:
        if head == None:
            return None
        sentinel = ListNode(0)
        sentinel.next = head
        first = sentinel
        second = sentinel
        while second != None:
            if second.next == None:
                first = first.next
                break
            else:
                first = first.next
                second = second.next.next
        return first

File: Leetcode_Java/middleNode.java
## 添加哨兵加快慢指针

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode middleNode(ListNode head) {
        if (head == null){
            return null;
        }
        ListNode sentinel = new ListNode(0);
        sentinel.next = head;
        ListNode slow = sentinel;
        ListNode fast = sentinel;
        while (fast != null){
            if (fast.next == null){
                slow = slow.next;
                break;
            }
            else{
                slow = slow.next;
                fast = fast.next.next;
            }
        }
        return slow;
    }
}
